sklearn/decomposition/_pca.py

```diff
@@ -533,33 +559,67 @@ def _fit_full(self, X, n_components):
                 )
         elif not 0 <= n_components <= min(n_samples, n_features):
             raise ValueError(
-                "n_components=%r must be between 0 and "
-                "min(n_samples, n_features)=%r with "
-                "svd_solver='full'" % (n_components, min(n_samples, n_features))
+                f"n_components={n_components} must be between 0 and "
+                f"min(n_samples, n_features)={min(n_samples, n_features)} with "
+                f"svd_solver={self._fit_svd_solver!r}"
             )
 
-        # Center data
         self.mean_ = xp.mean(X, axis=0)
-        X -= self.mean_
-
-        if not is_array_api_compliant:
-            # Use scipy.linalg with NumPy/SciPy inputs for the sake of not
-            # introducing unanticipated behavior changes. In the long run we
-            # could instead decide to always use xp.linalg.svd for all inputs,
-            # but that would make this code rely on numpy's SVD instead of
-            # scipy's. It's not 100% clear whether they use the same LAPACK
-            # solver by default though (assuming both are built against the
-            # same BLAS).
-            U, S, Vt = linalg.svd(X, full_matrices=False)
+        if self._fit_svd_solver == "full":
+            X_centered = xp.asarray(X, copy=True) if self.copy else X
+            X_centered -= self.mean_
+            x_is_centered = not self.copy
+
+            if not is_array_api_compliant:
+                # Use scipy.linalg with NumPy/SciPy inputs for the sake of not
+                # introducing unanticipated behavior changes. In the long run we
+                # could instead decide to always use xp.linalg.svd for all inputs,
+                # but that would make this code rely on numpy's SVD instead of
+                # scipy's. It's not 100% clear whether they use the same LAPACK
+                # solver by default though (assuming both are built against the
+                # same BLAS).
+                U, S, Vt = linalg.svd(X_centered, full_matrices=False)
+            else:
+                U, S, Vt = xp.linalg.svd(X_centered, full_matrices=False)
+            explained_variance_ = (S**2) / (n_samples - 1)
+
         else:
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
```

Comment:
No `assert` in production code, only tests. This should be covered by ruff, did we turn it off?

Reply:
Personally I like asserts: it communicates better the assumption we make in the code and make it more legible. But I can turn the assertion into an inline comment to communicate the intended purpose of the `else:` branch of the condition.

However, I would like to avoid the situation where we do:

```python
if self._fit_svd_solver in ("A", "B"):
    do something
elif self._fit_svd_solver == "C":
    do something else
else:
     raise ValueError(f"Invalid solver: {self._fit_svd_solver}")
```

because the `ValueError` branch can never be reached because it's already validated.

At the same time if we refactor other branches of the condition, inline comments run the risk of not being kept up-to-date, while assert statements would.

Reply:
Good point. In that case leave it as is. Maybe match when in python 3.10 will help with those in the future.

Reply:
Adding a message with the assertion could be nicer.